pallets/rc-migrator/src/crowdloan.rs

```diff
@@ -299,3 +299,73 @@ pub fn num_leases_to_ending_block<T: Config>(num_leases: u32) -> Result<BlockNum
 		.ok_or(())?;
 	Ok(last_period_end_block)
 }
+
+#[cfg(feature = "std")]
+impl<T: Config> crate::types::RcMigrationCheck for CrowdloanMigrator<T> {
+	/// Pre-migration payload for crowdloan data:
+	/// - `ParaId`: The parachain identifier
+	/// - Inner Vec contains contributions, where each tuple is:
+	///   - `BlockNumberFor<T>`: The block number at which this deposit can be unreserved
+	///   - `AccountId`: The depositor account
+	///   - `BalanceOf<T>`: The reserved amount
+	type RcPrePayload = Vec<(ParaId, Vec<(BlockNumberFor<T>, AccountIdOf<T>, BalanceOf<T>)>)>;
+
+	fn pre_check() -> Self::RcPrePayload {
+		// Iterate through all crowdloan funds and collect their contributions
+		pallet_crowdloan::Funds::<T>::iter()
+			.map(|(para_id, fund)| {
+				// For each fund, collect all contributions into a vector
+				let contributions =
+					pallet_crowdloan::Pallet::<T>::contribution_iterator(fund.fund_index)
+						.map(|(contributor, (amount, encoded_block_number))| {
+							// Decode the block number from bytes, defaulting to 0 if decoding fails
+							let block_number =
+								BlockNumberFor::<T>::decode(&mut &encoded_block_number[..])
+									.unwrap_or_default();
+
+							// Create a tuple of (block_number, contributor, amount)
+							// Convert amount to the expected balance type, defaulting to 0 if
+							// conversion fails
+							(block_number, contributor, amount.try_into().unwrap_or_default())
+						})
+						.collect();
+
+				// Return tuple of parachain ID and its contributions
+				(para_id, contributions)
+			})
+			.collect()
+	}
```

Comment:
We can still merge it, but in the future please dont nest the code so deeply. It is difficult to read.  
You can always set up a mutable vector and push into it with a for loop.